File: Java/String/Palindrome String.java
//https://www.geeksforgeeks.org/problems/palindrome-string0817/1
class Solution {
    boolean isPalindrome(String s) {
        // code here
        int st = 0, ed = s.length() - 1;

        while (st < ed) {
            while (st < ed && !Character.isLetterOrDigit(s.charAt(st))) {
                st++;
            }
            while (st < ed && !Character.isLetterOrDigit(s.charAt(ed))) {
                ed--;
            }

            if (Character.toLowerCase(s.charAt(st)) != Character.toLowerCase(s.charAt(ed))) {
                return false;
            }
            st++;
            ed--;
        }
        return true;
    }
};

File: Python/String/Palindrome String.py
//https://www.geeksforgeeks.org/problems/palindrome-string0817/1
class Solution:
    def isPalindrome(self, s: str) -> bool:
        st, ed = 0, len(s) - 1
        
        while st < ed:
            if not s[st].isalnum():
                st += 1
                continue
            if not s[ed].isalnum():
                ed -= 1
                continue
            
            if s[st].lower() != s[ed].lower():
                return False
            
            st += 1
            ed -= 1
            
        return True
